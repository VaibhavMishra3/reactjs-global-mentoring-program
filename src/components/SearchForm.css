.search-container {
  display: flex;
  align-items: center;
}

.search-input {
  width: 100%;
  padding: 10px;
  border-radius: 4px;
  border: 1px solid #555555;
  background-color: #232323;
  color: white;
}

.search-button {
  background-color: #f65261;
  color: white;
  border: none;
  padding: 10px 20px;
  cursor: pointer;
  border-radius: 4px;
  margin-left: 10px;
}

.search-button:hover {
  background-color: #e24350;
}
