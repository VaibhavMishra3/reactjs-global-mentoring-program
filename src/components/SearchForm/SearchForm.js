import React from 'react';
import { PropTypes } from 'prop-types';

const KEY_ENTER_CODE = 13;

class SearchForm extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      searchText: this.props.initialSearchText || ''
    };
  }

  render() {
    const searchText = this.state.searchText;

    return <div data-testid='search-form' className='container'>
      <label className='h4 form-label d-flex'>Find your movie</label>
      <div className='row'>
        <div className='col-9'>
          <input
            type='text'
            data-testid='search-input'
            value={searchText}
            onChange={this.handleInputChange}
            onKeyDown={this.handleInputKeyDown}
            placeholder={this.props.placeholderText}
            title={this.props.placeholderText}
            className='form-control'
          />
        </div>
        <div className='col-3'>
          <button
            type='button'
            data-testid='search-button'
            onClick={this.handleButtonClick}
            disabled={!searchText.length && 'disabled'}
            className='btn btn-primary d-flex'
          >
            {this.props.buttonText}
          </button>
        </div>
      </div>
    </div>
  }

  handleInputChange = (event) => {
    this.setState({
      searchText: event.target.value
    });
  }

  handleInputKeyDown = (event) => {
    if (event.keyCode === KEY_ENTER_CODE) {
      this.props.handleSearch(this.state.searchText);
    }
  }

  handleButtonClick = () => {
    this.props.handleSearch(this.state.searchText);
  }

}

SearchForm.propTypes = {
  initialSearchText: PropTypes.string.isRequired,
  placeholderText: PropTypes.string.isRequired,
  buttonText: PropTypes.string.isRequired,
  handleSearch: PropTypes.func.isRequired
};

export default SearchForm;
