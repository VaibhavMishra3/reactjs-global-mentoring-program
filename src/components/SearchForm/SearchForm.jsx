import React, { useState } from "react";
import "./SearchForm.css"; // Assuming you'll handle custom styling in this file

function SearchForm({ initialQuery, onSearch }) {
  const [query, setQuery] = useState(initialQuery);

  const handleSearch = () => {
    if (onSearch) onSearch(query);
  };

  const handleKeyPress = (event) => {
    if (event.key === "Enter") handleSearch();
  };

  return (
    <div className="search-container">
      <input
        type="text"
        className="search-input"
        placeholder="What do you want to watch?"
        value={query}
        onChange={(e) => setQuery(e.target.value)}
        onKeyPress={handleKeyPress}
      />
      <button className="search-button" onClick={handleSearch}>
        Search
      </button>
    </div>
  );
}

export default SearchForm;
