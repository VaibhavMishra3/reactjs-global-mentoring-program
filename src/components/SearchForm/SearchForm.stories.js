import SearchForm from './SearchForm';

export default {
  component: SearchForm
};

export const Default = {
  args: {
    initialSearchText: 'comedy',
    placeholderText: 'What do you want to watch?',
    buttonText: 'Search',
    handleSearch: (search) => console.log(`search: ${search}`),
  },
  parameters: {
  }
};
