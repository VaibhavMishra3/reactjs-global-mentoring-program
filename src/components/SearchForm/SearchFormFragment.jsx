import React from 'react'
import { PropTypes } from 'prop-types';
import { useRouter } from 'next/router';

import SearchForm from './SearchForm'
import { filterType } from '../../constants/types.js';
import { navigationByFilter } from '../../utils/navigation.js';

const SearchFormFragment = ({ filter, search, handleSearch }) => {
  const router = useRouter();

  return (
    <div>
      <span
        data-testid='add-movie'
        className='float-end'
        onClick={() => router.push(navigationByFilter('/new', filter))}
      >
        Add movie
      </span>
      <SearchForm
        initialSearchText={search}
        placeholderText='What do you want to watch?'
        buttonText='Search'
        handleSearch={handleSearch}
      />
    </div>
  )
}

SearchFormFragment.propTypes = {
  filter: filterType.isRequired,
  search: PropTypes.string.isRequired,
  handleSearch: PropTypes.func.isRequired
};

export default SearchFormFragment;
