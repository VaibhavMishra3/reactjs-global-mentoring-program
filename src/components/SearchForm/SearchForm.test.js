import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render, screen, fireEvent } from '@testing-library/react';

import SearchForm from './SearchForm';

const KEY_ENTER_CODE = 13;

describe('SearchForm', () => {

  afterEach(cleanup);

  it('should match snapshot', () => {
    const searchForm = renderer
      .create(
        <SearchForm
          initialSearchText='comedy'
          placeholderText='What do you want to watch?'
          buttonText='Search'
          handleSearch={console.log}
        />
      )
      .toJSON();

    expect(searchForm).toMatchSnapshot();
  });

  // Test that component renders an input with the value equal to initial value passed in props
  it('should render correctly', () => {
    render(
      <SearchForm
        initialSearchText='comedy'
        placeholderText='What do you want to watch?'
        buttonText='Search'
        handleSearch={console.log}
      />
    );

    const input = screen.getByTestId('search-input');
    expect(input).toBeInTheDocument();
    expect(input.tagName).toBe('INPUT');
    expect(input).toHaveAttribute('type', 'text');
    expect(input).toHaveValue('comedy');
    expect(input).toHaveAttribute('placeholder', 'What do you want to watch?');
    expect(input).toHaveAttribute('title', 'What do you want to watch?');

    const button = screen.getByTestId('search-button');
    expect(button).toBeInTheDocument();
    expect(button.tagName).toBe('BUTTON');
    expect(button).toHaveAttribute('type', 'button');
    expect(button).toBeEnabled();
    expect(button).toHaveTextContent('Search');
  });

  // Test that after typing to the input and a "click" event on the Submit button, the "onChange" prop is called with proper value
  it('should call the handler after typing to the input and clicking on the Submit button', () => {
    const handleSearchMock = jest.fn();

    render(
      <SearchForm
        initialSearchText='comedy'
        placeholderText='What do you want to watch?'
        buttonText='Search'
        handleSearch={handleSearchMock}
      />
    );

    const input = screen.getByTestId('search-input');
    const button = screen.getByTestId('search-button');

    fireEvent.change(input, { target: { value: 'comedy' } });
    fireEvent.click(button);

    expect(handleSearchMock).toHaveBeenCalledWith('comedy');
  });

  // Test that after typing to the input and pressing Enter key, the "onChange" prop is called with proper value
  it('should call the handler after typing to the input and pressing Enter key', () => {
    const handleSearchMock = jest.fn();

    render(
      <SearchForm
        initialSearchText='comedy'
        placeholderText='What do you want to watch?'
        buttonText='Search'
        handleSearch={handleSearchMock}
      />
    );

    const input = screen.getByTestId('search-input');

    fireEvent.change(input, { target: { value: 'comedy' } });
    fireEvent.keyDown(input, { keyCode: KEY_ENTER_CODE });

    expect(handleSearchMock).toHaveBeenCalledWith('comedy');
  });

});
