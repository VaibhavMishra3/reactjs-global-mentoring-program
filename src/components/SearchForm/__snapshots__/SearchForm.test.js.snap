// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm should match snapshot 1`] = `
<div
  className="container"
  data-testid="search-form"
>
  <label
    className="h4 form-label d-flex"
  >
    Find your movie
  </label>
  <div
    className="row"
  >
    <div
      className="col-9"
    >
      <input
        className="form-control"
        data-testid="search-input"
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="What do you want to watch?"
        title="What do you want to watch?"
        type="text"
        value="comedy"
      />
    </div>
    <div
      className="col-3"
    >
      <button
        className="btn btn-primary d-flex"
        data-testid="search-button"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        Search
      </button>
    </div>
  </div>
</div>
`;
